### ISSUE TITLE
LatexWalker infinite loop bug

### ISSUE THREAD

**Issue by Sinclert** [phfaist/pylatexenc#37]
There is a bug within the LatexWalker functionality, when a given LaTeX string incorrectly defines `\begin` or `\end` macros without the environment they refer to next to them.

### Issue

The bug produces this [LatexWalker.get_latex_nodes](https://github.com/phfaist/pylatexenc/blob/master/pylatexenc/latexwalker/__init__.py#L1739) function [while loop](https://github.com/phfaist/pylatexenc/blob/master/pylatexenc/latexwalker/__init__.py#L2169), to keep launching `LatexWalkerParseError` exceptions, while being forever ignored when the `self.tolerant_parsing` option is `True` (by default).

### How to reproduce?

```python
from pylatexenc.latex2text import LatexNodes2Text

content = r'''
    \begin
    \item {1.} Example text 1
    \item {2.} Example text 2
    \end
'''

conversor = LatexNodes2Text()
text = conversor.latex_to_text(content)
```

### Proposal
It is up to the maintainer of this package to consider this as a _"major"_ error and raise an exception to the user, or to log and ignore these special parsing errors.

My proposal is to consider this as a special type of parsing errors, logging a warning and continue with the parsing. I will create a Pull Request with the proposal.

**Comment by phfaist** [comment]
Thanks for spotting this bug! It seems to be a deeper issue with how `latexwalker` attempts to recover from errors in tolerant parsing mode. This is currently not handled particularly well, and now I've pushed a fix that should fix the issue along with some improvements on how error recovery is carried out. Thanks for your pull request, too. I decided to take the opportunity to carry out fixes that directly address the deeper issue instead, I'll comment about this directly on your PR.
